pycode/memilio-generation/memilio/generation/graph_visualization.py

```diff
@@ -0,0 +1,191 @@
+#
+# Authors: Daniel Richter
+#
+# Contact: Martin J. Kuehn <Martin.Kuehn@DLR.de>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#############################################################################
+"""
+@file graph_visualization.py
+@brief Class for plotting the abstract-syntax-tree in different formats. 
+"""
+import os
+from typing import TextIO
+from graphviz import Digraph
+from clang.cindex import Cursor
+from .ast import AST
+
+
+class Visualization:
+    """
+    Class for plotting the abstract-syntax-tree in different formats.
+    """
+    @staticmethod
+    def output_ast_terminal(ast: AST, cursor: Cursor) -> None:
+        """
+        Output the abstract syntax tree to terminal.
+
+        @param ast: ast object from AST class.
+        @param cursor: The current node of the AST as a cursor object from libclang.
+        """
+
+        _output_cursor_and_children_print(cursor, ast)
+
+        print(f'AST-terminal written.')
+
+    @staticmethod
+    def output_ast_png(cursor: Cursor, max_depth: int, ) -> None:
+        """
+        Output the abstract syntax tree to a .png. Set the starting node and the max depth.
+
+        @param cursor: The current node of the AST as a cursor object from libclang.
+        @param max_depth: Maximal depth the graph displays.
+        """
+
+        graph = Digraph(format='png')
+
+        _output_cursor_and_children_graphviz_digraph(
+            cursor, graph, max_depth, 0)
+
+        output_file = 'ast_graph'
+
+        graph.render(output_file, view=False)
+
+        output_path = os.path.abspath(f"{output_file}.png")
+        print(f'AST-png written to {output_path}')
+
+    @staticmethod
+    def output_ast_formatted(ast: AST, cursor: Cursor) -> None:
+        """
+        Output the abstract syntax tree to a file.
+
+        @param ast: ast object from AST class.
+        @param cursor: The current node of the AST as a cursor object from libclang.
+        """
+
+        with open('ast_formated.txt', 'w') as f:
+            _output_cursor_and_children_text(cursor, ast,  f)
+
+        print("AST-formated written to " + str(os.path.abspath(f.name)))
+
+
+def indent(level: int) -> str:
+    """
+    Create an indentation based on the level.
+    """
+    return '│   ' * level + '├── '
+
+
+def newline() -> str:
+    """
+    Create a new line.
+    """
+    return '\n'
+
+
+def _output_cursor_and_children_text(cursor: Cursor, ast: AST, f: TextIO, level: int = 0) -> int:
```

Comment:
This and the next function are mostly the same except the one line for the writer/printer. I would suggest to merge them into a single function and change the line `f.write(indent(level) + cursor_label + newline())` to a generic call to writer(level, cursor_label) and make writer one of the arguments of the function (instead of TextIO). Then each writer can be defined inside output_ast_terminal/_png